### DISCUSSION TITLE
About .NET Framework Wrapper

### DISCUSSION THREAD

**Discussion by touhiddnj** [githubrobbi/Ultra-Fast-File-Search#4] [Ideas]
I am using the UFFS COM file in the .NET project by grabbing the output which is not flexible for most .NET developers. If you provide a Wrapper/Package/Library for .Net Framework it will be very helpful for developers who are working on file searching.

Currently, I am using like below using C# language:

        private void runUltraFastFileSearch(object argument)
        {string args  = (string) argument;
            LogMessage("runUltraFastFileSearch initiated");
            string comFilePath = Path.GetDirectoryName(executablePath) + "\\uffs.com";
           

            ProcessStartInfo startInfo = new ProcessStartInfo
            {
                FileName = comFilePath,
                Arguments = args,
                UseShellExecute = false, // Set to false to run without the shell
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                CreateNoWindow = true, // Set to true to run without creating a window
            };

            using (Process process = new Process())
            {
                process.StartInfo = startInfo;
                process.Start();
                process.WaitForExit();

                string output = process.StandardOutput.ReadToEnd();
                string errors = process.StandardError.ReadToEnd();
                LogMessage(output);
                LogMessage(errors);

                // Handle output and errors as needed
            }
        }

**Comment by githubrobbi** [comment]
Hi @touhiddnj, sorry for the very late response!

A .NET wrapper would have been a nice addition, but I should mention that active development on this C++ codebase has ended.

The codebase was recently modernized (modular architecture, 95% monolith reduction, test infrastructure), but I've since rewritten the tool from scratch in Rust - same approach (direct NTFS MFT reading) but substantially faster and more scalable.

Your current approach of spawning the process and parsing output is actually solid and will continue to work. Tips to make it more flexible:

Use --columns= to control which fields you get
Use --sep= and --quotes= to simplify parsing
Use --out= for larger result sets
Pre-built binaries are available on the [Releases page](https://github.com/githubrobbi/Ultra-Fast-File-Search/releases/tag/v1.0.0).

If anyone wants to create a .NET wrapper or take over maintenance, PRs are welcome!

I wanted to leave a brief update on this repo.

I originally helped get a proper CLI wrapper in place. Earlier this month (**Jan 2026**), I also completed a significant refactor focused on codebase modernization:

### Codebase Modernization (2026-01)

- ~66 hours of refactoring work  
- ~95% monolith reduction (main file: **14,155 → 674 lines**)  
- Modular architecture (65+ headers across `src/cli`, `src/core`, `src/gui`, `src/index`, `src/io`, `src/search`, `src/util`)  
- Modern C++ usage (`[[nodiscard]]`, `noexcept`, `override`)  
- Test infrastructure added (33 unit tests using **doctest**)  
- Consistent `snake_case.hpp` naming  
- Self-contained headers with explicit dependencies  

That said, I’m no longer actively maintaining this codebase.

Since then, I rewrote the tool **from scratch in Rust**, implementing the same high-level approach: directly reading and processing the **NTFS MFT** rather than relying on OS filesystem commands. The Rust version is substantially faster and more scalable.

### Results (Rust implementation)

- Indexed **21,123,290 records** across 5 volumes  
  - 18,344,352 files  
  - 2,778,938 directories  
- **Read time:** 44.675s  
- **Total time:** 46.149s  

I’m happy to answer questions about the refactor or the Rust approach, but future maintenance will need a new owner.
